File: 1896-maximum-score-from-performing-multiplication-operations/1896-maximum-score-from-performing-multiplication-operations.java
class Solution {
    private int[] nums,multipliers;
    int find_max(int left,int mul,Integer[][] cache){
       if(mul==multipliers.length) return 0;
       if(cache[left][mul] != null) return cache[left][mul];
       int l = nums[left]*multipliers[mul]+find_max(left+1,mul+1,cache);
       int r = nums[(nums.length-1) - (mul-left)]*multipliers[mul]+find_max(left,mul+1,cache);
       cache[left][mul] = Math.max(l,r);
       return cache[left][mul];
    }
    public int maximumScore(int[] nums, int[] multipliers) {
       this.nums = nums;this.multipliers=multipliers;
       Integer[][] cache = new Integer[nums.length+1][multipliers.length+1];
       return  find_max(0,0,cache);
    }
}

File: 1896-maximum-score-from-performing-multiplication-operations/1896-maximum-score-from-performing-multiplication-operations.py
class Solution:
    def maximumScore(self, nums: List[int], multipliers: List[int]) -> int:
        
        # lets solve this with top-down dp

        # to perform any operation on nums, we need to consider both cases, whether if we consider left side number in nums or right side num in nums, we will pick optimal choice
        @functools.cache
        def recurse(left_side , nof_operation) -> int:
            
            #base case: (if we reach maximum number of operations)
            if (nof_operation == len(multipliers)) :
                return 0

            left_side_pick = nums[left_side] * multipliers[nof_operation] + recurse(left_side+1, nof_operation+1)

            # see here, we dont have any pointer to track on right side count, but we can find that by seeing that number of operations and number of left side operations. (if we subtract no. of operations performed so far - no. of operations performed on left side, we will get operations performed on right side), so if we subtract with last index with that, we will get the right index to work now.

            right_side_pick = nums[(len(nums)-1 - (nof_operation - left_side))] * multipliers[nof_operation] + recurse(left_side, nof_operation+1)

            return max(left_side_pick , right_side_pick)

        return recurse(0,0)
         


